Converted "row col-xx-xx" to "w-xx-xx" and moved hints outside the column

netwavebe opened the pull request:
A lot of xxxSettings.Edit.cshtml views had this markup in place:

```
<div class="row col-medium">
</div>
```

Since row has margin-right and margin-left -15px and col has padding-right and padding-left 15px an element should not contain both classes. Columns should always be child elements of a row.

I converted the markup to:

```
<div class="row">
    <div class="col-medium">
    </div>
</div>
```

I also moved any hints placed inside a column so that they are not restricted by the width of the input-element.

furthermore I nomalized all hint inputs to `col-medium`.

Random screenshot of Numeric field settings:

Before:

![image](https://user-images.githubusercontent.com/3008547/97578210-c7364480-19f0-11eb-91ff-f73e7bdf95e2.png)

After:

![image](https://user-images.githubusercontent.com/3008547/97578237-cdc4bc00-19f0-11eb-827d-bda2a10d5d6e.png)

Right margin is now always the same.

Skrypt commented:
Yeah it's me that did this. While semantically it seems wrong you generally have nothing directly in the `<div class="row">` so it's adding a `div` for nothing. But for code clarity I agree on the changes.

Skrypt reviewed (approved):
Yeah, it's me that did this. While semantically it seems wrong you generally have nothing directly in the <div class="row"> that is not a column `<div class="col-x-x">` so it's adding a div for nothing. But for code clarity I agree on the changes.

Skrypt commented on src/Templates/OrchardCore.ProjectTemplates/content/OrchardCore.Templates.Cms.Module/Views/MyTestPartSettings.Edit.cshtml:
```diff
@@ -1,9 +1,11 @@
-        <input asp-for="MySetting" class="form-control" />
-        <span class="hint">@T["The setting."]</span>
+    <div class="row">
+        <div class="col-large">
```
col-large ???

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentTypes/Views/ContentPartSettings.Edit.cshtml:
```diff
@@ -2,20 +2,24 @@
-        <input asp-for="Description" class="form-control" />
-        <span class="hint">@T["Optionally provide a description for this part."]</span>
+    <div class="row">
+        <div class="col-large">
```
col-large ???

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentTypes/Views/Admin/AddFieldTo.cshtml:
```diff
@@ -12,19 +12,23 @@
-            <input asp-for="Name" class="form-control" />
-            <span class="hint">@T["Technical name of the field."]</span>
+        <div class="row">
+            <div class="col-small">
```
col-sm

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentTypes/Views/Admin/AddFieldTo.cshtml:
```diff
@@ -12,19 +12,23 @@
-            <input asp-for="DisplayName" class="form-control" autofocus />
-            <span class="hint">@T["Name of the field as it will be displayed in screens."]</span>
+        <div class="row">
+            <div class="col-medium">
```
col-md

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Views/YoutubeFieldSetting.Edit.cshtml:
```diff
@@ -13,22 +13,33 @@
-        <textarea asp-for="Hint" rows="2" class="form-control"></textarea>
-        <span class="hint">@T["The hint text to display for this field on the editor."]</span>
+    <div class="row">
+        <div class="col-medium">
```
col-md

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Views/TimeFieldSettings.Edit.cshtml:
```diff
@@ -9,9 +9,11 @@
-        <textarea asp-for="Hint" rows="2" class="form-control"></textarea>
-        <span class="hint">@T["The hint text to display for this field on the editor."]</span>
+    <div class="row">
+        <div class="col-medium">
```
col-md

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Views/TextFieldSettings.Edit.cshtml:
```diff
@@ -9,9 +9,11 @@
-        <textarea asp-for="Hint" rows="2" class="form-control"></textarea>
-        <span class="hint">@T["The hint text to display for this field on the editor."]</span>
+    <div class="row">
+        <div class="col-medium">
```
col-md

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Views/NumericFieldSettings.Edit.cshtml:
```diff
@@ -17,49 +17,61 @@
-        <input asp-for="DefaultValue" class="form-control" min="@minimum" max="@maximum" step="@stepAttribute" type="number" />
-        <span class="hint">@T["The default value. (optional)"]</span>
+    <div class="row">
+        <div class="col-small">
```
col-sm

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Views/NumericFieldSettings.Edit.cshtml:
```diff
@@ -17,49 +17,61 @@
-        <input asp-for="Maximum" class="form-control" />@*step="@step" type="number"*@
-        <span class="hint">@T["The maximum value allowed. (optional)"]</span>
+    <div class="row">
+        <div class="col-small">
```
col-sm

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Views/NumericFieldSettings.Edit.cshtml:
```diff
@@ -17,49 +17,61 @@
-        <input asp-for="Placeholder" type="text" class="form-control" />
-        <span class="hint">@T["A hint to display when the input is empty. (optional)"]</span>
+    <div class="row">
+        <div class="col-medium">
```
col-md

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Views/NumericFieldSettings.Edit.cshtml:
```diff
@@ -17,49 +17,61 @@
-        <textarea asp-for="Hint" rows="2" class="form-control"></textarea>
-        <span class="hint">@T["The hint text to display for this field on the editor."]</span>
+    <div class="row">
+        <div class="col-medium">
```
col-md

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Views/LocalizationSetContentPickerFieldSettings.Edit.cshtml:
```diff
@@ -24,9 +24,11 @@
-        <textarea asp-for="Hint" rows="2" class="form-control"></textarea>
-        <span class="hint">@T["The hint text to display for this field on the editor."]</span>
+    <div class="row">
+        <div class="col-medium">
```
col-md

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Views/DateTimeFieldSettings.Edit.cshtml:
```diff
@@ -9,9 +9,11 @@
-        <textarea asp-for="Hint" rows="2" class="form-control"></textarea>
-        <span class="hint">@T["The hint text to display for this field on the editor."]</span>
+    <div class="row">
+        <div class="col-medium">
```
col-md

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Views/DateFieldSettings.Edit.cshtml:
```diff
@@ -9,9 +9,11 @@
-        <textarea asp-for="Hint" rows="2" class="form-control"></textarea>
-        <span class="hint">@T["The hint text to display for this field on the editor."]</span>
+    <div class="row">
+        <div class="col-medium">
```
col-md

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Views/ContentPickerFieldSettings.Edit.cshtml:
```diff
@@ -46,19 +46,23 @@
-        <textarea asp-for="Hint" rows="2" class="form-control"></textarea>
-        <span class="hint">@T["The hint text to display for this field on the editor."]</span>
+    <div class="row">
+        <div class="col-medium">
```
col-md

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Views/BooleanFieldSettings.Edit.cshtml:
```diff
@@ -1,19 +1,23 @@
-        <input asp-for="Label" class="form-control" />
-        <span class="hint">@T["The text associated to the checkbox."]</span>
+    <div class="row">
+        <div class="col-small">
```
col-sm

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Views/BooleanFieldSettings.Edit.cshtml:
```diff
@@ -1,19 +1,23 @@
-        <textarea asp-for="Hint" rows="2" class="form-control"></textarea>
-        <span class="hint">@T["The hint text to display for this field on the editor."]</span>
+    <div class="row">
+        <div class="col-medium">
```
col-md

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Views/NumericFieldSettings.Edit.cshtml:
```diff
@@ -17,49 +17,61 @@
-        <input asp-for="Scale" class="form-control" min="0" max="5" step="1" type="number" />
-        <span class="hint">@T["The number of digits after the decimal point."]</span>
+    <div class="row">
+        <div class="col-small">
```
col-sm

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Views/NumericFieldSettings.Edit.cshtml:
```diff
@@ -17,49 +17,61 @@
-        <input asp-for="Minimum" class="form-control" />@*step="@step" type="number"*@
-        <span class="hint">@T["The minimum value allowed. (optional)"]</span>
+    <div class="row">
+        <div class="col-small">
```
col-sm

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentTypes/Views/Admin/EditPart.cshtml:
```diff
@@ -15,11 +15,13 @@
-            <input asp-for="Name" class="form-control" readonly />
-            <span class="hint">@T["Technical name of the part."]</span>
+        <div class="row">
+            <div class="col-small">
```
col-sm

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentTypes/Views/ContentTypeSettings.Edit.cshtml:
```diff
@@ -41,9 +41,11 @@
-        <input asp-for="Stereotype" type="text" class="form-control">
-        <span class="hint">@T["(Optional) The stereotype of the content type. e.g., Widget, MenuItem, ..."]</span>
+    <div class="row">
+        <div class="col-medium">
```
col-md

Skrypt commented on src/OrchardCore.Modules/OrchardCore.Taxonomies/Views/TaxonomyFieldSettings.Edit.cshtml:
```diff
@@ -11,22 +11,26 @@
-        <textarea asp-for="Hint" rows="2" class="form-control"></textarea>
-        <span class="hint">@T["The hint text to display for this field on the editor."]</span>
+    <div class="row">
+        <div class="col-medium">
```
col-md

netwavebe commented:
@Skrypt Thanks for your feedback Jasmin!

I would not say 'it's adding a div for nothing' as this change would fix the right margin issues on small screen.

I noticed `col-md` and `col-medium` were both used, as well as `w-md-50`. I too wanted to stick to one style, but there are differences between them. Look at this screenshot:

![image](https://user-images.githubusercontent.com/3008547/97675524-c30d3400-1a8f-11eb-8634-5b974f88448c.png)

if `col-md` is not followed by a number, it's full width.

The origin of `col-medium` can be found here:

https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Themes/TheAdmin/Assets/scss/modules/admin/_grid.scss#L12

So basically it's a shortcode for `col-12 col-sm-12 col-md-10 col-lg-8 col-xl-6`.

Replacing `col-medium` with `col-md` or `col-md-6` or `w-md-50` would all result in a very different layout. As I did not want to change layout, I reused whatever was already in place (except for hints inputs, I changed those to `col-medium`).

The same goes for `col-small` and `col-large`: can't replace them without impacting layout.

hishamco commented:
@deanmarcussen how this differ from what you did previously?

netwavebe commented:
@hishamco What changeset is that?

hishamco commented:
AFAIK I started a PR organize and align the elements and use the suitable css classes then @deanmarcussen replaces my PR and fix them, I'm just asking him if we miss something or why I see thing come again ;)

netwavebe commented:
@hishamco I see. I'm sure Dean will jump in soon.

I do think this needs to be fixed, as the current markup is against Bootstrap's guidelines.

From https://getbootstrap.com/docs/4.5/layout/grid/:

- Rows are wrappers for columns. Each column has horizontal padding (called a gutter) for controlling the space between them. This padding is then counteracted on the rows with negative margins. This way, all the content in your columns is visually aligned down the left side.
- In a grid layout, content must be placed within columns and only columns may be immediate children of rows.

Using `row col-x-x` on the same element causes layout issues as shown in the first screenshot (that element has both padding and a negative margin).

I'm a sucker for clean UI's I'm afraid... :wink:

deanmarcussen commented:
Agree with @Skrypt requests.

The col-medium style classes should be phased out, they are weird, and don't fit the rest of ui.
I reestablished them after breaking them (to fix the ui short term), but with the hope they would be phased out afterwards.

The grid padding is the problem with the sizing classes `w-md-50` in my opinion @Skrypt 

Using just a sizing class means any other fields in the display inside a grid, i.e. two cols of 6, means that padding is out by 15px, when compared to something sized with a `w`

@hishamco your pr missed a few things, the crucial one at the time, was what @netwavebe has done here which was moving the hints outside of the sized classes.

Opinions only

hishamco commented:
> @hishamco your pr missed a few things,

I didn't mean anything about PR, just I thought your PR fix this

> was moving the hints outside of the sized classes

I see

netwavebe commented:
Hi @deanmarcussen, thanks for your insight.

The difference between `row > col-md-6` versus `w-md-50` is bothering me for some time now, because of the 15px margin difference you mention. Some screens look weird/messy because of this. I am in favor of using `row > col` instead of `w-xx-xx` because they are more flexible (multiple columns, 1/3 width + 2/3 width, etc). Even if this means an extra div-element.

With this in mind, I'd rather not change this PR to `w-xx-xx` style but keep the `row > col` classes in place.

If the `col-small`, `col-medium`, ... classes are deprecated, I suggest I replace them to what they currently stand for (but simplified):

`col-medium` (= `col-12 col-sm-12 col-md-10 col-lg-8 col-xl-6`) becomes `col-12 col-md-10 col-lg-8 col-xl-6`
`col-large` (= `col-12 col-sm-12 col-md-12 col-lg-12 col-xl-9`) becomes `col-12 col-xl-9`

This way there is no change in layout whatsoever. If needed, they can later be simplified in another PR.

The `col-small`, `col-medium`, ... classes can be removed later.

Any thoughts @deanmarcussen and @Skrypt ?

Skrypt commented:
I still think these forms are in no need to use a grid. This is a vertical form which can be sized with the w-x-x class. The margin you are talking about only applies when you use a `row` class which in this case would not be necessary. If you want to use a grid per form element because the components themselves are separated in columns then it makes sense. Here, we have a form that has virtually no need to use columns since they are all one column. This is why normally we should be using w-x-x class instead to reduce the HTML markup required to style these.

I agree we should phase out the helper classes col-medium, col-small, col-large as we should never extend base classes like this to apply grouped styles together everywhere, it just make the .css file grow in size and make things more complicated for theming extensibility.

So basically, if it's using `col-12 col-sm-12 col-md-10 col-lg-8 col-xl-6` it should be `w-100 w-sm-100 w-md-75 w-xl-50`

netwavebe commented:
Mixing `w-xx-xx` and `row > col-xx-xx` on the same form (field settings or otherwise) will always result in 15px difference gap because of the 'gutter' the `row > col-xx-xx` inherently introduces. To have a truly clean UI, we need to get rid of that difference gap. This mean we should focus on one style, stick with it, and loose the other.

I think we can both agree on this, right?

While it's true that most field settings are one-column forms, there are exceptions. Look at the LinkField settings editor for example: https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.ContentFields/Views/LinkFieldSettings.Edit.cshtml#L5

It shows url and text data in two columns. This particular field can't be refactored using `w-xx-xx` because the 'gutter' between the two columns is essential.

This isn't limited to field settings, it applies to content type editors as well. I've created several custom parts with multiple columns (a recent example: latitude and longitude values, in two columns).

By opting for `w-xx-xx`, we would effectively enforce people to stick to just one column. Or, if they do use `row > col-xx-xx` in their field setting or part, it would look out of place because of the metioned 15px difference gap.

So imho, `row > col-xx-xx` is much more flexible. It doesn't put a constrain on what people can build without looking out of place. It also gives people more fine-grained control over how everything looks (12 columns vs 4).

netwavebe commented:
In regards to the container you mentioned: class `ta-content` had a padding of 15px and thus acts as a valid container for Bootstrap. Nesting containers is not needed.

See https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Themes/TheAdmin/Assets/scss/modules/_layout.scss#L32

Skrypt commented:
I do not totally agree 100% with all these statements. I believe that we should be able to use both methods in a single form.

Let's use this form as an example which @deanmarcussen has made.

https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.Media/Views/MediaProfiles/Edit.cshtml

Let's add a unique form-group to this view and see how it would work if we'd use a grid on a single element.

Here first form-group currently is : 

``` cshtml
    <div class="form-group " asp-validation-class-for="Name">
        <div class="w-100 w-md-50">
            <label asp-for="Name">@T["Name"]</label>
            <input asp-for="Name" class="form-control" autofocus />
            <span asp-validation-for="Name"></span>
        </div>
        <div class="w-100">
            <span class="hint">@T["The name of the media profile."]</span>
        </div>
    </div>
```

If we would use a grid : 

``` cshtml
    <div class="form-group " asp-validation-class-for="Name">
        <div class="row">
            <div class="col col-md-6">
                <label asp-for="Name">@T["Name"]</label>
                <input asp-for="Name" class="form-control" autofocus />
                <span asp-validation-for="Name"></span>
            </div>
            <div class="col-12">
                <span class="hint">@T["The name of the media profile."]</span>
            </div>
        </div>
    </div>
```

> It also gives people more fine-grained control over how everything looks (12 columns vs 4).

You can add $sizes this is not an issue. If this is what we need then let's add the same sizes than what the grid offers. Right now, I did not change the default $sizes because there was no real need so far.

netwavebe commented:
100% agreed, to be able to use both methods in a single form would be perfect.

This PR involves field settings. These forms can be checked and tested. If there's a need for 2 columns, like the LinkField, a `px-3` can be added to 'fix' the 15px gap difference between them.

But what about `AddFieldTo.cshtml` and `EditPart.cshtml` that are also in this PR? Would they need the `px-3` fix? They would for a LinkField, but not for a NumericField or TextField. So the gap difference would still exist.

How can this be fixed?

PS: Adding more sizes is not something I would want to do. Imho these classes are no better than the `col-medium` classes @deanmarcussen wanted to remove. Frontend builders know stuff like `w-50` and `col-md-50`, but `w-md-50` and `col-medium` is unknown stuff.

Skrypt commented:
Yes, we need to document it but for that matter we need to have a base set of SASS mixins that could be shareable througout our modules maybe by sharing them through npm. Things are in dev and I'm not much into documentation since I'm not the best to write it. We still need to implement things better to allow extending our base admin theme. Right now each modules have their own SASS files and we can't share the TheAdmin theme mixins, func, vars with our module which is blocking me from moving on with our own SASS mixins documentation.

netwavebe commented:
It's a pity you commented on my last remark but did not share your insight on how to fix the `AddFieldTo.cshtml` and `EditPart.cshtml` issue.

I decided to convert my changes to `w-xx-xx` because I feel like I've invested too much of my free time in what could have been an easy change, but turned out to be an endless discussion instead.

At least it'll solve the right gap on small screens.

netwavebe commented:
What I mean is this:

![image](https://user-images.githubusercontent.com/3008547/97798664-5ed1a800-1c28-11eb-80a7-7da544e7e5cc.png)

This is the difference between `w-xx-xx` and using `row > col-xx-xx`. This can be fixed by adding `px-3` in `AddFieldTo.cshtml` and `EditField.cshtml` but then the difference will show up in other fields. This also occurs in various parts at this moment.

I guess I focus too much on clean UI's... All in all it's not important enough to start a discussion like we had.

As said, an update with `w-xx-xx` will come asap.

That LinkField does look odd, I agree. I checked: the TextField is next to the Required field by design. I will update that in another PR, okay? Don't want to do too much in this one.

netwavebe commented:
What I mean is this:

![image](https://user-images.githubusercontent.com/3008547/97798664-5ed1a800-1c28-11eb-80a7-7da544e7e5cc.png)

This is the difference between `w-xx-xx` and using `row > col-xx-xx`. This can be fixed by adding `px-3` in `AddFieldTo.cshtml` and `EditField.cshtml` but then the difference will show up in other fields. This also occurs in various parts at this moment.

I guess I focus too much on clean UI's... All in all it's not important enough to start a discussion like we had.

As said, an update with `w-xx-xx` will come asap.

That LinkField does look odd, I agree. I checked: the TextField is next to the Required field by design. I will update that in another PR, okay? Don't want to do too much in this one.

netwavebe commented:
Mmm, not sure why my comment suddenly appears before yours instead of chronologically?

Skrypt commented:
Okay leaving you decide but you got my explanation ! 😉 
Let me know when you are done and I will merge it.

Skrypt commented:
![screencapture-localhost-5000-Admin-ContentParts-Edit-AliasPart-2020-11-01-03_23_43](https://user-images.githubusercontent.com/3228637/97798192-c9b3bc80-1bf1-11eb-970f-d21dcf6bf030.png)
![screencapture-localhost-5000-Admin-ContentTypes-AddFieldsTo-Article-2020-11-01-03_21_59](https://user-images.githubusercontent.com/3228637/97798193-ca4c5300-1bf1-11eb-8e5e-259e3812a8ee.png)

These 2 forms are one column layout so I would use w-breakpoint-size. Not sure where you see there would be a potential issue?

![screencapture-localhost-5000-Admin-ContentParts-Article-Fields-terw-Edit-2020-11-01-03_33_22](https://user-images.githubusercontent.com/3228637/97798339-409d8500-1bf3-11eb-8b7b-c7c89b1adc53.png)

LinkField looks like a 2 column layout with a mixed pattern which works fine but elements are not positioned correctly.

NumericField and TextField without looking seems to me like a one column layout.
If you see something with those 2 that I've not seen please show me.

Sorry for the long explanations, I think we obviously miss documentation and/or do not understand some of your points. Happens ...

netwavebe commented:
I converted `row > col-xx-xx` to `w-xx-xx`. for `col-name` I used the following conversion:

`col-small` = `col-12 col-sm-9 col-md-6 col-lg-3 col-xl-3` = `w-sm-75 w-md-50 w-lg-25`
`col-medium` = `col-12 col-sm-12 col-md-10 col-lg-8 col-xl-6` = `w-md-75 w-xl-50`
`col-large` = `col-12 col-sm-12 col-md-12 col-lg-12 col-xl-9` = `w-xl-75`

Remember I changed all hint input-elements to col-medium previously so now they all have `w-md-75 w-xl-50` classes.

I added EditField.cshtml to the list of changed files to match the width of the 'Technical name' field in 'AddFieldTo.cshtml'.

netwavebe commented on src/OrchardCore.Modules/OrchardCore.Taxonomies/Views/TaxonomyFieldSettings.Edit.cshtml:
```diff
@@ -11,15 +11,15 @@
-<div class="row">
-    <div class="form-group col-sm-6">
+<div class="form-group">
+    <div class="w-md-50">
```
col-sm-6 was too small imho, changed it to w-md-50.

netwavebe commented:
@Skrypt I think everything is converted to `w-xx-xx` correctlty now.

Do you have the time to take a quick look at the file changes?

Skrypt commented:
Will do now. Need to sleep also sometimes ... 😄
